vertx-web/src/main/java/io/vertx/ext/web/impl/Utils.java

```diff
@@ -40,93 +42,254 @@
+   *
+   * double slash // will be converted to single slash and the path will always start with slash.
    *
-   * @param path
-   *          The original path
-   * @param result
-   *          The result of unescaping the escape sequence (and removing dangerous constructs)
-   * @param i
-   *          The index of path where the escape sequence begins
-   * @return The index of path where the escape sequence ends
-   * @throws UnsupportedEncodingException
-   *           If the escape sequence does not represent a valid UTF-8 string
+   * @param pathname raw path
+   * @return normalized path
    */
-  private static int processEscapeSequence(String path, StringBuilder result, int i) throws UnsupportedEncodingException {
-    Buffer buf = Buffer.buffer(2);
-    do {
-      if (i >= path.length() - 2) {
-        throw new IllegalArgumentException("Invalid position for escape character: " + i);
-      }
-      int unescaped = Integer.parseInt(path.substring(i + 1, i + 3), 16);
-      if (unescaped < 0) {
-        throw new IllegalArgumentException("Invalid escape sequence: " + path.substring(i, i + 3));
+  public static String normalizePath(String pathname) {
+    // add trailing slash if not set
+    if (pathname == null || pathname.length() == 0) {
+      return "/";
+    }
+
+    StringBuilder ibuf = new StringBuilder(pathname.length() + 1);
+
+    // Not standard!!!
+    if (pathname.charAt(0) != '/') {
+      ibuf.append('/');
+    }
+
+    ibuf.append(pathname);
+    int i = 0;
+
+    while (i < ibuf.length()) {
+      // decode unreserved chars described in
+      // http://tools.ietf.org/html/rfc3986#section-2.4
+      if (ibuf.charAt(i) == '%') {
+        decodeUnreserved(ibuf, i);
       }
-      buf.appendByte((byte) unescaped);
-      i += 3;
-    } while (i < path.length() && path.charAt(i) == '%');
-
-    String escapedSeq = new String(buf.getBytes(), StandardCharsets.UTF_8);
-
-    for (int j = 0; j < escapedSeq.length(); j++) {
-      char c = escapedSeq.charAt(j);
-      if (c == '/') {
-        if (j == 0 || result.charAt(result.length() - 1) != '/')
-          result.append(c);
-      } else if (c == '.') {
-        if (j == 0 || result.charAt(result.length() - 1) != '.')
-          result.append(c);
-        else
-          result.deleteCharAt(result.length() - 1);
+
+      i++;
+    }
+
+    // remove dots as described in
+    // http://tools.ietf.org/html/rfc3986#section-5.2.4
+    return removeDots(ibuf);
+  }
+
+  /**
+   * Removed dots as per <a href="http://tools.ietf.org/html/rfc3986#section-5.2.4>rfc3986</a>.
+   *
+   * There are 2 extra transformations that are not part of the spec but kept for backwards compatibility:
+   *
+   * double slash // will be converted to single slash and the path will always start with slash.
+   *
+   * @param ibuf raw path
+   * @return normalized path
+   */
+  public static String removeDots(CharSequence ibuf) {
```

Comment:
it should be private since "removeDots" is the naming used for the algorithm/pseudocode in the RFC. I'll document with that in mind